linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -4683,4 +4683,191 @@ private void flush() {
       }
     };
   }
+
+  /**
+   * Merge Union Enumerable.
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   *
+   * @param sources input enumerables (must be already sorted)
+   * @param sortKeySelector sort key selector
+   * @param sortComparator sort comparator to decide the next item
+   * @param all whether duplicates will be considered or not
+   * @param equalityComparer {@link EqualityComparer} to control duplicates,
+   *                         only used if {@code all} is {@code false}
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  public static <TSource, @Nullable TKey> Enumerable<TSource> mergeUnion(
+      List<Enumerable<TSource>> sources,
+      Function1<TSource, @Nullable TKey> sortKeySelector,
+      Comparator<TKey> sortComparator,
+      boolean all,
+      EqualityComparer<TSource> equalityComparer) {
+    return new AbstractEnumerable<TSource>() {
+      @Override public Enumerator<TSource> enumerator() {
+        return new MergeUnionEnumerator<>(
+            sources,
+            sortKeySelector,
+            sortComparator,
+            all,
+            equalityComparer);
+      }
+    };
+  }
+
+  /**
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  @SuppressWarnings("unchecked")
+  private static class MergeUnionEnumerator<TSource, @Nullable TKey>
+      implements Enumerator<TSource> {
+    private final List<Enumerator<TSource>> inputs;
+    private final TSource[] currentInputsValues;
+    private final Function1<TSource, @Nullable TKey> sortKeySelector;
+    private final Comparator<@Nullable TKey> sortComparator;
+    private TSource currentValue;
+    private int activeInputs;
+
+    // Set to control duplicates, only used if "all" is false
+    private final @Nullable Set<Wrapped<TSource>> processed;
+    private final @Nullable Function1<TSource, Wrapped<TSource>> wrapper;
+
+    private static final Object NOT_INIT = new Object();
+    private static final Object FINISHED = new Object();
+
+    @SuppressWarnings("method.invocation.invalid")
```

Comment:
Am I right this is only for `initEnumerators`?
If so, it would be great to move the suppression to the relevant method only + add `@UnderInitialization this` parameter to `initEnumerators`

Reply:
Yes, this is about `initEnumerators`:
```
error: [method.invocation.invalid] call to initEnumerators() not allowed on the given receiver.
      this.initEnumerators();
                          ^
  found   : @UnderInitialization(org.apache.calcite.linq4j.EnumerableDefaults.MergeUnionEnumerator.class) @NonNull MergeUnionEnumerator
  required: @Initialized @NonNull MergeUnionEnumerator
```
Sorry, I did not understand your problem,  how should I fix it?

Reply:
In general,`suppressions` should be targeted. If you add `SuppressWarnings` to the full constructor, it might allow unexpected warnings to creep in.

In other words, suppressions should be confined to their minimal scope.

For instance,

```java
    @RequiresNonNull("inputs")
    @SuppressWarnings("method.invocation.invalid")
    private void initEnumerators(
        EnumerableDefaults.@UnknownInitialization MergeUnionEnumerator<TSource, TKey> this
    ) {
      for (int i = 0; i < this.inputs.size(); i++) {
        this.moveEnumerator(i);
      }
    }
```

Reply:
Done